### PR TITLE
[stable] GHA: Sync main.yml with DMD repo's

### PR DESCRIPTION
**Pull request by kinke**
Mainly to work around the macOS 13 issue.

### CODE REVIEW

**Review by thewilsonator — APPROVED** [review]

File: .github/workflows/main.yml [hunk 1]
```diff
@@ -80,6 +83,13 @@ jobs:
         with:
           arch: ${{ env.MODEL == '64' && 'x64' || 'x86' }}
 
+      # NOTE: Linker ICEs with Xcode 15.0.1 (default version on macos-13)
+      # * https://issues.dlang.org/show_bug.cgi?id=24407
+      # Remove this step if the default gets changed to 15.1 in actions/runner-images.
+      - name: 'macOS 13: Switch to Xcode v15.1'
+        if: matrix.os == 'macos-13'
+        run: sudo xcode-select -switch /Applications/Xcode_15.1.app
+
```
Comment (ibuclaw) [inline .github/workflows/main.yml:92]: Do we trigger this in this repo?

Comment (kinke) [reply .github/workflows/main.yml:92]: Sure; the steps are the same as for the DMD CI jobs. Both master/stable branches of Phobos are currently red, e.g., https://github.com/dlang/phobos/actions/runs/8118656250/job/22193306501.

Comment (ibuclaw) [reply .github/workflows/main.yml:92]: The only test I'm aware of that triggers the Xcode 15.0 bug is one that instantiates >32k templates (See the test case in the linked issue).

So unless phobos does that too, I think we can safely omit this step. :-)

Comment (kinke) [reply .github/workflows/main.yml:92]: Phobos CI here runs the same tests as DMD. So if the problem is a test in `compiler/tests`, then it's tested by Phobos CI here too.

Comment (ibuclaw) [reply .github/workflows/main.yml:92]: Hmm, that really shouldn't be the case, but OK.

**Pull request merged**